devicetypes/krlaframboise/ecolink-chime-siren.src/ecolink-chime-siren.groovy

```diff
@@ -0,0 +1,694 @@
+				cmds << secureCmd(zwave.configurationV1.configurationSet(parameterNumber: param.num, size: param.size, scaledConfigurationValue: param.value))
+				cmds << secureCmd(zwave.configurationV1.configurationGet(parameterNumber: param.num))
+			}
+		}
+	}
+	sendCommands(cmds)
+}
+
+def ping() {
+	logDebug "ping()..."
+	return [ batteryGetCmd() ]
+}
+
+def setChimeVolume(chimeVolume) {
+	sendEventIfNew("chimeVolume", chimeVolume, "%")
+}
+
+def setChimeSound(chimeSound) {
+	sendEventIfNew("chimeSound", chimeSound)
+}
+
+def chime() {
+	logDebug "chime()..."
+
+	int volume = safeToInt(device.currentValue("chimeVolume"), 50)
+	int sound = safeToInt(device.currentValue("chimeSound"), 1)
+
+	state.pendingAction = "chime"
+	playSoundAtVolume(sound, volume)
+}
+
+def setSoundVolume(soundVolume) {
+	sendEventIfNew("soundVolume", soundVolume, "%")
+}
+
+def playSound(soundNumber) {
+	logDebug "playSound(${soundNumber})..."
+
+	int volume = safeToInt(device.currentValue("soundVolume"), 25)
+	int sound = safeToInt(soundNumber, 1)
+
+	state.pendingAction = soundNumber
+	playSoundAtVolume(sound, volume)
+}
+
+def setSirenVolume(sirenVolume) {
+	sendEventIfNew("sirenVolume", sirenVolume, "%")
+}
+
+def setSirenSound(sirenSound) {
+	sendEventIfNew("sirenSound", sirenSound)
+}
+
+def both() {
+	siren()
+}
+
+def strobe() {
+	siren()
+}
+
+def siren() {
+	logDebug "siren()..."
+
+	int volume = safeToInt(device.currentValue("sirenVolume"), 100)
+	int sound = safeToInt(device.currentValue("sirenSound"), 13)
+
+	state.pendingAction = "siren"
+	playSoundAtVolume(sound, volume)
+}
+
+void playSoundAtVolume(soundNumber, volume) {
+	logDebug "playSoundAtVolume(${soundNumber}, ${volume})..."
+
+	Map sound = getSound(soundNumber)
+	state.lastSound = sound
+	
+	logDebug "Playing '${sound.name}' at ${volume}%..."
+
+	List<String> cmds = [
+		soundSwitchConfigSetCmd(volume, 1)
+	]
+
+	if (sound.indicatorID) {
+		cmds << secureCmd(zwave.indicatorV1.indicatorSet(value: sound.indicatorID))
+	} else {
+		cmds << soundSwitchTonePlaySetCmd(soundNumber)
+	}
+
+	cmds << soundSwitchTonePlayGetCmd()
+
+	sendCommands(cmds, 100)
+}
+
+def on() {
+	chime()
+}
+
+def off() {
+	logDebug "off()..."
+	return delayBetween([
+		soundSwitchTonePlaySetCmd(0),
+		soundSwitchTonePlayGetCmd()
+	])
+}
+
+def refresh() {
+	logDebug "refresh()..."
+	sendCommands([
+		batteryGetCmd(),
+		secureCmd(zwave.versionV1.versionGet()),
+		soundSwitchTonePlayGetCmd()
+	])
+}
+
+void sendCommands(List<String> cmds, Integer delay=1000) {
+	if (cmds) {
+		def actions = []
+		cmds.each {
+			actions << new physicalgraph.device.HubAction(it)
+		}
+		sendHubCommand(actions, delay)
+	}
+}
+
+String batteryGetCmd() {
+	return secureCmd(zwave.batteryV1.batteryGet())
+}
+
+String secureCmd(cmd) {
+	try {
+		if (isSecurityEnabled()) {
+			return zwave.securityV1.securityMessageEncapsulation().encapsulate(cmd).format()
+		} else {
+			return cmd.format()
+		}
+	} catch (ex) {
+		return cmd.format()
```

Comment:
catch block is for handling errors, why do You  `return cmd.format() `here ?
Is any exception thrown for a virtual device ?

Reply:
The line below throws an exception for a virtual device, even though it shouldn't because ? is a null check.  

```
boolean isSecurityEnabled() {
	return zwaveInfo?.zw?.contains("s")
}
```

I used it as a virtual device a lot while building the presentation, but it's unlikely end users will do that so let me know if you want it removed...

Reply:
zwaveInfo might not be defined for a virtual device,  that's why a runtime exception was thrown in isSecurityEnabled().
`} catch (ex) {` block might be removed because such errors don't occur for real devices.